mllib/src/main/scala/org/apache/spark/mllib/clustering/AffinityPropagation.scala

```diff
@@ -0,0 +1,475 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.clustering
+
+import scala.collection.mutable
+
+import org.apache.spark.{Logging, SparkException}
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.api.java.JavaRDD
+import org.apache.spark.graphx._
+import org.apache.spark.graphx.impl.GraphImpl
+import org.apache.spark.rdd.RDD
+
+/**
+ * :: Experimental ::
+ *
+ * Model produced by [[AffinityPropagation]].
+ *
+ * @param id cluster id.
+ * @param exemplar cluster exemplar.
+ * @param members cluster members.
+ */
+@Experimental
+case class AffinityPropagationCluster(val id: Long, val exemplar: Long, val members: Array[Long])
+
+/**
+ * :: Experimental ::
+ *
+ * Model produced by [[AffinityPropagation]].
+ *
+ * @param clusters the clusters of AffinityPropagation clustering results.
+ */
+@Experimental
+class AffinityPropagationModel(
+    val clusters: RDD[AffinityPropagationCluster]) extends Serializable {
+
+  /**
+   * Set the number of clusters
+   */
+  lazy val getK: Long = clusters.count()
+ 
+  /**
+   * Find the cluster the given vertex belongs
+   * @param vertexID vertex id.
+   * @return a [[Array]] that contains vertex ids in the same cluster of given vertexID. If
+   *         the given vertex doesn't belong to any cluster, return null.
+   */
+  def findCluster(vertexID: Long): Array[Long] = {
```

Comment:
I have the same concern @mengxr has. Can that particular cluster contains lots of vertexes that run out of memory?  This can return `RDD[Long]`, but with data structure of `(vertex id, cluster id)`, it seems it requires two passes.